File: Machine Learning A-Z/Part 4 - Clustering/Section 25 - Hierarchical Clustering/hc.R
dataset = read.csv('Mall_Customers.csv')
x <- dataset[4:5]

dendogram=hclust(d=dist(x,method = "euclidean"),method = 'ward.D')
plot(dendogram,main=paste('Dendogram'),xlab='hclust',ylab='Distance')

hc=hclust(d=dist(x,method = "euclidean"),method = 'ward.D')
y_hc=cutree(hc,5)


library(cluster)
clusplot(x,y_hc,lines=0,shade=TRUE,color=TRUE,plotchar =FALSE,span = TRUE,main=paste('client clusters'),xlab='Annual Income',ylab='Spending Score',labels = 2)

File: Machine Learning A-Z/Part 4 - Clustering/Section 25 - Hierarchical Clustering/hc.py
# Data Preprocessing Template

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Mall_Customers.csv')
X = dataset.iloc[:, [3,4]].values

import scipy.cluster.hierarchy as sch
dendogram=sch.dendrogram(sch.linkage(X,method='ward'))
plt.title('Dendogram')
plt.xlabel('customers')
plt.ylabel('Distance')
plt.show()

from sklearn.cluster import AgglomerativeClustering
hc=AgglomerativeClustering(n_clusters=5,affinity='euclidean',linkage='ward')
y_hc=hc.fit_predict(X)


colors=['red','blue','green','cyan','magenta']
titles=['Careful','Standard','Target','Careless','Sensible']
for i in range(0,5):
    plt.scatter(X[y_hc == i,0],X[y_hc == i, 1],s=100,c=colors[i],label=titles[i])

plt.title('HC')
plt.xlabel('Income')
plt.ylabel('Score')
plt.legend()
plt.show()